File: backend/src/haan/tools/code_tools.py
"""
Code analysis tools for Haan.ai agents.

Provides 2 tools mirroring the TypeScript code-tools.ts:
- CodeAnalyzeTool:     Analyze code structure (imports, exports, symbols)
- CodeFindSymbolTool:  Search for symbol definitions across the codebase
"""

from __future__ import annotations

import os
import re
import subprocess
from pathlib import Path
from typing import Any

from crewai.tools import BaseTool as CrewAIBaseTool

from haan.utils.path_utils import get_project_root, relative_path, safe_path


# Directories to skip during code analysis
SKIP_DIRS = {".git", "node_modules", "dist", "__pycache__", ".venv", "venv", ".tox"}

# Source file extensions to include in analysis
SOURCE_EXTENSIONS = {
    ".ts", ".tsx", ".js", ".jsx",
    ".py", ".go", ".rs", ".java",
    ".c", ".cpp", ".h", ".hpp",
    ".rb", ".php", ".swift", ".kt",
}


class CodeAnalyzeTool(CrewAIBaseTool):
    """Analyze code structure of a file or project."""

    name: str = "code-analyze"
    description: str = (
        "Analyze code structure of a file or project directory. "
        "For directories: reads package.json/pyproject.toml and lists source files. "
        "For files: finds imports, exports, classes, and functions."
    )

    def _run(self, path: str) -> str:
        """
        Analyze code structure.

        Args:
            path: File or directory path relative to project root.

        Returns:
            Structured analysis output.
        """
        try:
            target = Path(safe_path(path))
            if not target.exists():
                return f"Error: Path not found: {path}"

            if target.is_dir():
                return self._analyze_directory(target)
            else:
                return self._analyze_file(target)
        except Exception as e:
            return f"Error: Analysis failed: {e}"

    def _analyze_directory(self, dir_path: Path) -> str:
        """Analyze a project directory structure."""
        sections: list[str] = []

        # Check for package.json (Node.js)
        pkg_path = dir_path / "package.json"
        if pkg_path.exists():
            import json
            try:
                pkg = json.loads(pkg_path.read_text())
                sections.append("=== package.json ===")
                sections.append(f"Name: {pkg.get('name', 'unnamed')}")
                sections.append(f"Version: {pkg.get('version', 'unversioned')}")
                if pkg.get("description"):
                    sections.append(f"Description: {pkg['description']}")
                if pkg.get("main"):
                    sections.append(f"Main: {pkg['main']}")
                if pkg.get("scripts"):
                    sections.append(f"Scripts: {', '.join(pkg['scripts'].keys())}")
                deps = pkg.get("dependencies", {})
                if deps:
                    sections.append(f"Dependencies ({len(deps)}): {', '.join(deps.keys())}")
                dev_deps = pkg.get("devDependencies", {})
                if dev_deps:
                    sections.append(
                        f"DevDependencies ({len(dev_deps)}): {', '.join(dev_deps.keys())}"
                    )
            except (json.JSONDecodeError, KeyError):
                sections.append("package.json: failed to parse")

        # Check for pyproject.toml (Python)
        pyproject_path = dir_path / "pyproject.toml"
        if pyproject_path.exists():
            sections.append("\n=== pyproject.toml ===")
            sections.append("Python project detected")

        # Check for tsconfig.json (TypeScript)
        if (dir_path / "tsconfig.json").exists():
            sections.append("\n=== tsconfig.json ===")
            sections.append("TypeScript project detected")

        # List source files
        src_dir = dir_path / "src"
        search_dir = src_dir if src_dir.exists() else dir_path
        sections.append(f"\n=== Source Structure ({relative_path(str(search_dir))}) ===")

        files = self._collect_source_files(search_dir)
        root = get_project_root()
        for f in files[:50]:
            sections.append(str(f).replace(root + "/", ""))
        if len(files) > 50:
            sections.append(f"... and {len(files) - 50} more files")

        return "\n".join(sections)

    def _analyze_file(self, file_path: Path) -> str:
        """Analyze a single source file for imports, exports, and symbols."""
        content = file_path.read_text(encoding="utf-8", errors="replace")
        lines = content.split("\n")
        sections: list[str] = []
        rel = relative_path(str(file_path))

        sections.append(f"=== {rel} ===")
        sections.append(f"Lines: {len(lines)}")

        # Find imports
        imports: list[str] = []
        for line in lines:
            # JavaScript/TypeScript imports
            m = re.match(r"^import\s+.*from\s+['\"](.+)['\"]", line)
            if m:
                imports.append(m.group(1))
                continue
            # Python imports
            m = re.match(r"^(?:from\s+(\S+)\s+import|import\s+(\S+))", line)
            if m:
                imports.append(m.group(1) or m.group(2))
                continue
            # CommonJS require
            m = re.search(r"require\s*\(\s*['\"](.+)['\"]\s*\)", line)
            if m:
                imports.append(m.group(1))

        if imports:
            sections.append(f"\nImports ({len(imports)}):")
            for imp in imports:
                sections.append(f"  {imp}")

        # Find exports (JS/TS)
        exports: list[str] = []
        for line in lines:
            m = re.match(
                r"^export\s+(?:default\s+)?"
                r"(class|function|const|let|var|interface|type|enum|abstract\s+class)\s+(\w+)",
                line,
            )
            if m:
                exports.append(f"{m.group(1)} {m.group(2)}")

        if exports:
            sections.append(f"\nExports ({len(exports)}):")
            for exp in exports:
                sections.append(f"  {exp}")

        # Find all symbol definitions
        symbols: list[str] = []
        for i, line in enumerate(lines):
            # Classes
            m = re.match(r"\s*(?:abstract\s+)?class\s+(\w+)", line)
            if m:
                symbols.append(f"class {m.group(1)} (line {i + 1})")
            # Functions (JS/TS/Python)
            m = re.match(r"\s*(?:async\s+)?(?:function|def)\s+(\w+)", line)
            if m:
                symbols.append(f"function {m.group(1)} (line {i + 1})")

        if symbols:
            sections.append(f"\nSymbols ({len(symbols)}):")
            for sym in symbols:
                sections.append(f"  {sym}")

        return "\n".join(sections)

    @staticmethod
    def _collect_source_files(directory: Path) -> list[Path]:
        """Recursively collect source files, skipping excluded directories."""
        results: list[Path] = []

        def _walk(d: Path) -> None:
            try:
                entries = sorted(d.iterdir())
            except PermissionError:
                return
            for entry in entries:
                if entry.name.startswith(".") or entry.name in SKIP_DIRS:
                    continue
                if entry.is_dir():
                    _walk(entry)
                elif entry.suffix in SOURCE_EXTENSIONS:
                    results.append(entry)

        _walk(directory)
        return results


class CodeFindSymbolTool(CrewAIBaseTool):
    """Search for symbol definitions across the codebase."""

    name: str = "code-find-symbol"
    description: str = (
        "Search for symbol definitions (class, function, const, interface, type, def, func) "
        "across the codebase. Returns matching file paths and line numbers."
    )

    def _run(self, symbol: str, path: str = "src") -> str:
        """
        Search for symbol definitions.

        Args:
            symbol: Symbol name to search for.
            path: Directory to search in (relative to project root).

        Returns:
            Matching definitions with file paths and line numbers.
        """
        try:
            search_path = safe_path(path)
            root = get_project_root()

            # Build a pattern that matches common definition forms across languages
            patterns = [
                rf"(class|interface|type|enum|function|const|let|var|abstract class)\s+{re.escape(symbol)}\b",
                rf"def\s+{re.escape(symbol)}\s*\(",   # Python
                rf"func\s+{re.escape(symbol)}\s*\(",   # Go
                rf"fn\s+{re.escape(symbol)}\s*[(<]",    # Rust
            ]
            combined = "|".join(patterns)

            # Try ripgrep first, fall back to grep
            try:
                subprocess.run(["which", "rg"], capture_output=True, check=True)
                cmd = [
                    "rg", "--no-heading", "--line-number",
                    "--max-count", "50", "-e", combined, search_path,
                ]
            except subprocess.CalledProcessError:
                cmd = [
                    "grep", "-rn", "--max-count=50",
                    "-E", combined, search_path,
                ]

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=10,
                cwd=root,
            )

            output = result.stdout.strip()
            if not output:
                return f"No definitions found for symbol: {symbol}"

            # Relativize paths
            relativized = output.replace(root + "/", "")
            return relativized

        except subprocess.TimeoutExpired:
            return "Error: Symbol search timed out."
        except Exception as e:
            return f"Error: Symbol search failed: {e}"

File: src/tools/code-tools.ts
import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';
import { BaseTool, ToolResult } from './types.js';
import { safePath, getProjectRoot, relativePath } from '../utils/path-utils.js';

export class CodeAnalyzeTool extends BaseTool {
  name = 'code-analyze';
  description = 'Analyze code structure of a file or project. Reads package.json, finds imports, and lists exports.';
  parameters = {
    type: 'object' as const,
    properties: {
      path: { type: 'string', description: 'File or directory path relative to project root' },
    },
    required: ['path'],
  };

  async execute(params: Record<string, unknown>): Promise<ToolResult> {
    try {
      const targetPath = safePath(params.path as string);
      const stat = fs.statSync(targetPath);

      if (stat.isDirectory()) {
        return this.analyzeDirectory(targetPath);
      } else {
        return this.analyzeFile(targetPath);
      }
    } catch (err) {
      return this.error(`Analysis failed: ${(err as Error).message}`);
    }
  }

  private analyzeDirectory(dirPath: string): ToolResult {
    const sections: string[] = [];

    // Check for package.json
    const pkgPath = path.join(dirPath, 'package.json');
    if (fs.existsSync(pkgPath)) {
      try {
        const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'));
        sections.push('=== package.json ===');
        sections.push(`Name: ${pkg.name || 'unnamed'}`);
        sections.push(`Version: ${pkg.version || 'unversioned'}`);
        if (pkg.description) sections.push(`Description: ${pkg.description}`);
        if (pkg.main) sections.push(`Main: ${pkg.main}`);
        if (pkg.scripts) {
          sections.push(`Scripts: ${Object.keys(pkg.scripts).join(', ')}`);
        }
        if (pkg.dependencies) {
          sections.push(`Dependencies (${Object.keys(pkg.dependencies).length}): ${Object.keys(pkg.dependencies).join(', ')}`);
        }
        if (pkg.devDependencies) {
          sections.push(`DevDependencies (${Object.keys(pkg.devDependencies).length}): ${Object.keys(pkg.devDependencies).join(', ')}`);
        }
      } catch {
        sections.push('package.json: failed to parse');
      }
    }

    // Check for tsconfig.json
    const tsconfigPath = path.join(dirPath, 'tsconfig.json');
    if (fs.existsSync(tsconfigPath)) {
      sections.push('\n=== tsconfig.json ===');
      sections.push('TypeScript project detected');
    }

    // List source files
    const srcDir = path.join(dirPath, 'src');
    if (fs.existsSync(srcDir)) {
      sections.push('\n=== Source Structure ===');
      const files = this.collectFiles(srcDir, ['.ts', '.tsx', '.js', '.jsx', '.py', '.go', '.rs']);
      const root = getProjectRoot();
      for (const f of files.slice(0, 50)) {
        sections.push(f.replace(root + '/', ''));
      }
      if (files.length > 50) {
        sections.push(`... and ${files.length - 50} more files`);
      }
    }

    return this.success(sections.join('\n'));
  }

  private analyzeFile(filePath: string): ToolResult {
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n');
    const sections: string[] = [];
    const rel = relativePath(filePath);

    sections.push(`=== ${rel} ===`);
    sections.push(`Lines: ${lines.length}`);

    // Find imports
    const imports: string[] = [];
    for (const line of lines) {
      const importMatch = line.match(/^import\s+.*from\s+['"](.+)['"]/);
      if (importMatch) {
        imports.push(importMatch[1]);
        continue;
      }
      const requireMatch = line.match(/require\s*\(\s*['"](.+)['"]\s*\)/);
      if (requireMatch) {
        imports.push(requireMatch[1]);
      }
    }
    if (imports.length > 0) {
      sections.push(`\nImports (${imports.length}):`);
      for (const imp of imports) {
        sections.push(`  ${imp}`);
      }
    }

    // Find exports
    const exports: string[] = [];
    for (const line of lines) {
      const exportMatch = line.match(/^export\s+(default\s+)?(class|function|const|let|var|interface|type|enum|abstract\s+class)\s+(\w+)/);
      if (exportMatch) {
        const kind = exportMatch[2];
        const name = exportMatch[3];
        exports.push(`${kind} ${name}`);
      }
    }
    if (exports.length > 0) {
      sections.push(`\nExports (${exports.length}):`);
      for (const exp of exports) {
        sections.push(`  ${exp}`);
      }
    }

    // Find classes and functions (non-exported too)
    const symbols: string[] = [];
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const classMatch = line.match(/(?:abstract\s+)?class\s+(\w+)/);
      if (classMatch) {
        symbols.push(`class ${classMatch[1]} (line ${i + 1})`);
      }
      const funcMatch = line.match(/(?:async\s+)?function\s+(\w+)/);
      if (funcMatch) {
        symbols.push(`function ${funcMatch[1]} (line ${i + 1})`);
      }
    }
    if (symbols.length > 0) {
      sections.push(`\nSymbols (${symbols.length}):`);
      for (const sym of symbols) {
        sections.push(`  ${sym}`);
      }
    }

    return this.success(sections.join('\n'));
  }

  private collectFiles(dir: string, extensions: string[]): string[] {
    const results: string[] = [];
    const walk = (d: string) => {
      let entries: fs.Dirent[];
      try {
        entries = fs.readdirSync(d, { withFileTypes: true });
      } catch {
        return;
      }
      for (const entry of entries) {
        if (entry.name.startsWith('.') || entry.name === 'node_modules' || entry.name === 'dist') continue;
        const full = path.join(d, entry.name);
        if (entry.isDirectory()) {
          walk(full);
        } else {
          const ext = path.extname(entry.name);
          if (extensions.includes(ext)) {
            results.push(full);
          }
        }
      }
    };
    walk(dir);
    return results;
  }
}

export class CodeFindSymbolTool extends BaseTool {
  name = 'code-find-symbol';
  description = 'Search for symbol definitions (class, function, const, interface, type) across the codebase.';
  parameters = {
    type: 'object' as const,
    properties: {
      symbol: { type: 'string', description: 'Symbol name to search for' },
      path: { type: 'string', description: 'Directory to search in (relative to project root, defaults to "src")' },
    },
    required: ['symbol'],
  };

  async execute(params: Record<string, unknown>): Promise<ToolResult> {
    try {
      const symbol = params.symbol as string;
      const searchPath = safePath((params.path as string) || 'src');
      const root = getProjectRoot();

      // Pattern matches common definition forms
      const patterns = [
        `(class|interface|type|enum|function|const|let|var|abstract class)\\s+${symbol}\\b`,
        `def\\s+${symbol}\\s*\\(`,     // Python
        `func\\s+${symbol}\\s*\\(`,     // Go
        `fn\\s+${symbol}\\s*[(<]`,      // Rust
      ];
      const combinedPattern = patterns.join('|');

      let cmd: string;
      try {
        execSync('which rg', { stdio: 'pipe' });
        cmd = `rg --no-heading --line-number --max-count 50 -e '${combinedPattern}' '${searchPath}'`;
      } catch {
        cmd = `grep -rn --max-count=50 -E '${combinedPattern}' '${searchPath}'`;
      }

      try {
        const output = execSync(cmd, {
          encoding: 'utf-8',
          timeout: 10000,
          maxBuffer: 512 * 1024,
          cwd: root,
        }).trim();

        if (!output) {
          return this.success(`No definitions found for symbol: ${symbol}`);
        }

        const relativized = output
          .split('\n')
          .map((line) => line.replace(root + '/', ''))
          .join('\n');

        return this.success(relativized);
      } catch (execErr: unknown) {
        const exitErr = execErr as { status?: number };
        if (exitErr.status === 1) {
          return this.success(`No definitions found for symbol: ${symbol}`);
        }
        throw execErr;
      }
    } catch (err) {
      return this.error(`Symbol search failed: ${(err as Error).message}`);
    }
  }
}
